drafts/swift-async-await-cycle.md

```diff
@@ -0,0 +1,122 @@
+
+![Visualisation of the retain cycle](https://cdn.hashnode.com/res/hashnode/image/upload/v1724685726758/0b1QCVlR4.png?auto=format)
+
+In the real project, the `Task` iterated over an `AyncSequence` that acted as a message queue (that’s why we needed it in the first place) and the `sendMessage` method implemented a timeout mechanism that reported a failure after 750ms. Since the connection of the preceding test was force-terminated, the message-sending never finished, keeping the `MessageLayer`  alive, which in turn resulted in the timeout being triggered while the next test was already executing.
+
+Now let’s talk about solutions.
+
+## Solution #1: Cancel the task 
+The simplest option is to cancel the task from the outside before removing the reference to the `MessageLayer`. By adding `messageLayer?.task?.cancel()` before (4.) above we achieve the desired behaviour:
+```text
+Sending: 1
+Deinit
+End of test
+```
+
+No delayed deallocation. Unfortunately, this requires users of the `MessageLayer` to remember cancelling the task. This is not very intuitive and requires internal knowledge of the `MessageLayer` — not great.
+ 
+## Solution #2: Remove self-reference
+Another option is to make `sendMessages` static (or even put it outside the class). This removes the (implicit) reference to `self`  and allows the object to be freely deallocated (in turn cancelling the `Task`).
+
+This requires us to pass in all `MessageLayer` state the method needs, now that it has no access to `self` anymore. We just need to be careful to not cause a classical retain cycle when doing so.
+
+Here’s the code, still resulting in the same output as in solution #1:
+```swift
+class MessageLayer {
+    private var task: Task<Void, Never>? // private now
+
+    init() {
+        task = Task { // Be careful to not capture self here strongly
+            try? await Self.sendMessages(2) // Self instead of self
+        }
+    }
+
+    private static func sendMessages(_ messages: Int) async throws { // static
+        for i in 1...messages {
+            print("Sending: \(i)")
+            try await Task.sleep(for: .seconds(0.5))
+        }
+    }
+
+    deinit {
+        print("Deinit")
+        task?.cancel()
+    }
+}
+
+var messageLayer: MessageLayer? = MessageLayer()
+messageLayer = nil
+print("End of test")
+```
+
+The nice thing is that the background task management is completely self contained and consumers of the `MessageLayer` can just use it without a second thought.
+
+## Wrap Up
+As you can see, it is quite simple to inadvertently extend the lifetime of objects with long-running async functions. We implemented solution #2 to solve our issue successfully. 
+
+We hope we saved you some hours of debugging and you enjoyed diving into this issue as much as we did. If you can avoid spawning a new `Task` all together, you never run into this problem. That’s obviously preferred. However, in case you cannot get around it, you now have a solution at hand.
```

Comment:
```suggestion
We hope we saved you some hours of debugging and that you enjoyed diving into this issue as much as we did. If you can avoid spawning a new `Task` altogether, you never run into this problem. That’s obviously preferred. However, in case you cannot get around it, you now have a solution at hand.
```